swig/python2/python_maps2.i

```diff
@@ -0,0 +1,132 @@
+
+//typemap for short maxlen strings
+%typemap(in, numinputs=0) (char *outputString, int maxlen) {
+  $2=256;
+  $1=(char *)malloc(256);
+}
+
+%typemap(argout) (char *outputString, int maxlen) {
+  PyObject *str=PyString_FromString($1);
+  $result = SWIG_Python_AppendOutput($result, str);
+}
+
+%typemap(freearg) (char *outputString, int maxlen) {
+   if ($1) free($1);
+}
+
+
+//typemap for large string output with a length return in C
+%typemap(in, numinputs=0) (char *outputString, int maxStringlen, int *actualLength) {
+  $3=&($2);
+}
+
+%typemap(freearg) (char *outputString, int maxStringlen, int *actualLength) {
+   if ($1) free($1);
+}
+
+%typemap(check)(char *outputString, int maxStringlen, int *actualLength) {
+    $2=helicsSubscriptionGetValueSize(arg1)+2;
+    $1 = (char *) malloc($2);
+}
+
+%typemap(argout) (char *outputString, int maxStringlen, int *actualLength) {
+  PyObject *o2=PyString_FromString($1);
+  $result = SWIG_Python_AppendOutput($result, o2);
+}
+
+
+//typemap for the input arguments
+%typemap(in) (int argc, const char *const *argv) {
+  /* Check if is a list */
+  if (PyList_Check($input)) {
+    int i;
+    $1 = PyList_Size($input);
+    $2 = (char **) malloc(($1+1)*sizeof(char *));
+    for (i = 0; i < $1; i++) {
+      PyObject *o = PyList_GetItem($input,i);
+      if (PyString_Check(o))
+	$2[i] = PyString_AsString(PyList_GetItem($input,i));
+      else {
+	PyErr_SetString(PyExc_TypeError,"list must contain strings");
+	free($2);
+	return NULL;
+      }
+    }
+    $2[i] = 0;
+  } else {
+    PyErr_SetString(PyExc_TypeError,"not a list");
+    return NULL;
+  }
+}
+
+%typemap(freearg) (int argc, const char *const *argv) {
+  free((char *) $2);
+}
+
+// typemap for vector input functions
+%typemap(in) (const double *vectorInput, int vectorlength) {
+  int i;
+  if (!PyList_Check($input)) {
+    PyErr_SetString(PyExc_ValueError,"Expected a list");
+    return NULL;
+  }
+  $2=PyList_Size($input);
+  $1 = (double *) malloc($2*sizeof(double));
+
+  for (i = 0; i < $2; i++) {
+    PyObject *o = PyList_GetItem($input,i);
+    if (PyFloat_Check(o)) {
+      $1[i] = PyFloat_AsDouble(o);
+	}else if (PyInt_Check(o))
+	{
+		$1[i] = (double)(PyInt_AsLong(o));
+	} else {
+      PyErr_SetString(PyExc_ValueError,"List elements must be numbers");
+      free($1);
+      return NULL;
+    }
+  }
+}
+
+%typemap(argout) (const double *vectorInput, int vectorlength)
+{
+}
+
+%typemap(freearg) (const double *vectorInput, int vectorlength) {
+   if ($1) free($1);
+}
+
+// typemap for vector output functions
+
+%typemap(arginit) (double data[], int maxlen, int *actualSize) {
+  $1=(double *)(NULL);
```

Comment:
Question: Why is $1 set to NULL?

Reply:
Mostly to prevent compiler warnings for uninitialized values.  in swig inputs $1, $2, etc are references to the actual objects that go into the C call which in this case is double *.  We actually set it to something in the (in) typemap but that can error, and if that happens then if $1 was not set to NULL it could attempt to free it, which would be bad.